### PR TITLE
Cancelling a peer recovery on the source can leak a primary permit

### PR DESCRIPTION
**Pull request by bleskes**
The code in `SourceRecoveryHandler` runs under a `CancellableThreads` instance in order to allow long running operations to be interrupted when the recovery is cancelled. Sadly if this happens at just the wrong moment while acquiring a permit from the primary, that primary can be leaked and never be freed.

Note that this is slightly better than it sounds - we only cancel recoveries on the source side if the primary shard itself is closed.

Relates to https://github.com/elastic/elasticsearch/pull/30316

### CODE REVIEW

**Comment by bleskes** [comment]
Thx @jasontedor for the brainstorm session around a clean solution

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -227,17 +229,38 @@ private boolean isTargetSameHistory() {
...
+                if (primary.isPrimaryMode() == false) {
+                    throw new IndexShardRelocatedException(primary.shardId());
                 }
                 runnable.run();
+            } finally {
+                // just in case we got an exception (likely interrupted) while waiting for the get
+                permit.whenComplete((r, e) -> {
+                    if (r != null) {
+                        r.close();
+                    }
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Should we do anything with the exception, like log it?

**Review by jasontedor — APPROVED** [review]
I left a question. I trust your judgement.

LGTM.

**Review by ywelsch — APPROVED** [review]
LGTM. I don't understand why this needs to be a blocker, though.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 1 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: good question. I think in all likely hood the exception will already bubble up through the get. I can log it in trace (and wire in a logger).

**Comment by bleskes** [comment]
> I don't understand why this needs to be a blocker, though.

@ywelsch it's good question - the path I found isn't really (as the primary shard is being shutdown anyway) but since it's a disaster waiting to happen and super tricky to debug, I want it in. I don't think it will slow down anything because we'll be respining due to #30316

**Pull request merged**